File: lectures/lecture-06-why-agents-overreach-and-under-finish/exercises/exercise-02-completion-gate/solution/python/main.py
"""completion-gate exercise, Python solution.

Audits a feature list's `passing` claims against the evidence rule (the
recorded command is the feature's own verification command and the
recorded run passed), checks the WIP limit, and says through its exit
code whether the next feature may be activated. SPEC.md pins the rules,
the strings, and the verdict precedence.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

WIP_LIMIT = 1


def gate(features: list[dict]) -> tuple[dict, int]:
    claims = []
    unbacked: list[str] = []
    for feature in features:
        if feature["status"] != "passing":
            continue
        evidence = feature.get("evidence")
        command = feature["verification"]
        if evidence is None:
            ok, detail = False, "no evidence recorded"
        elif evidence["command"] != command:
            ok, detail = False, (
                f"evidence names a different command ({evidence['command']}, not {command})"
            )
        elif not evidence["observed"].startswith("exit 0"):
            ok, detail = False, f"evidence records a failing run ({evidence['observed']})"
        else:
            ok, detail = True, f"verified: {command} reported exit 0"
        claims.append({"id": feature["id"], "evidence_ok": ok, "detail": detail})
        if not ok:
            unbacked.append(feature["id"])
    in_progress = [feature["id"] for feature in features if feature["status"] == "in-progress"]
    if len(in_progress) > WIP_LIMIT:
        verdict, code = "wip-exceeded", 1
    elif unbacked:
        verdict, code = "unbacked-claims", 1
    else:
        verdict, code = "sound", 0
    report = {
        "claims": claims,
        "may_activate": verdict == "sound" and not in_progress,
        "unbacked": unbacked,
        "verdict": verdict,
        "wip": {"in_progress": in_progress, "limit": WIP_LIMIT},
    }
    return report, code


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: main.py <feature_list.json>", file=sys.stderr)
        return 2
    try:
        data = json.loads(Path(argv[1]).read_text(encoding="utf-8"))
    except OSError as error:
        print(f"error: cannot read feature list: {error}", file=sys.stderr)
        return 2
    report, code = gate(data["features"])
    print(json.dumps(report, indent=2))
    return code


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: lectures/lecture-06-why-agents-overreach-and-under-finish/exercises/exercise-02-completion-gate/solution/typescript/main.ts
// completion-gate exercise, TypeScript solution.
//
// Audits a feature list's `passing` claims against the evidence rule (the
// recorded command is the feature's own verification command and the
// recorded run passed), checks the WIP limit, and says through its exit
// code whether the next feature may be activated. SPEC.md pins the rules,
// the strings, and the verdict precedence.

import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

const WIP_LIMIT = 1;

interface Evidence {
  command: string;
  observed: string;
  date: string;
}

interface Feature {
  id: string;
  status: string;
  verification: string;
  evidence?: Evidence;
}

export function gate(features: Feature[]): [object, number] {
  const claims = [];
  const unbacked: string[] = [];
  for (const feature of features) {
    if (feature.status !== "passing") {
      continue;
    }
    const evidence = feature.evidence ?? null;
    const command = feature.verification;
    let ok: boolean;
    let detail: string;
    if (command.trim() === "") {
      ok = false;
      detail = "the feature declares no verification command";
    } else if (evidence === null) {
      ok = false;
      detail = "no evidence recorded";
    } else if (evidence.command !== command) {
      ok = false;
      detail = `evidence names a different command (${evidence.command}, not ${command})`;
    } else if (!evidence.observed.startsWith("exit 0")) {
      ok = false;
      detail = `evidence records a failing run (${evidence.observed})`;
    } else {
      ok = true;
      detail = `verified: ${command} reported exit 0`;
    }
    claims.push({ id: feature.id, evidence_ok: ok, detail });
    if (!ok) {
      unbacked.push(feature.id);
    }
  }
  const inProgress = features.filter((f) => f.status === "in-progress").map((f) => f.id);
  let verdict: string;
  let code: number;
  if (inProgress.length > WIP_LIMIT) {
    verdict = "wip-exceeded";
    code = 1;
  } else if (unbacked.length > 0) {
    verdict = "unbacked-claims";
    code = 1;
  } else {
    verdict = "sound";
    code = 0;
  }
  const report = {
    claims,
    may_activate: verdict === "sound" && inProgress.length === 0,
    unbacked,
    verdict,
    wip: { in_progress: inProgress, limit: WIP_LIMIT },
  };
  return [report, code];
}

function main(argv: readonly string[]): number {
  const path = argv[2];
  if (!path || argv.length !== 3) {
    console.error("usage: main.ts <feature_list.json>");
    return 2;
  }
  let data: { features: Feature[] };
  try {
    data = JSON.parse(readFileSync(path, "utf8"));
  } catch (error) {
    console.error(`error: cannot read feature list: ${(error as Error).message}`);
    return 2;
  }
  const [report, code] = gate(data.features);
  process.stdout.write(JSON.stringify(report, null, 2) + "\n");
  return code;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.exit(main(process.argv));
}
